src/Numeric/Units/Dimensional/Dynamic.hs

```diff
@@ -101,19 +101,23 @@ instance Num a => Monoid (AnyQuantity a) where
 -- not be a 'Quantity' of any 'Dimension' whatsoever, but in exchange it gains instances
 -- for the common numeric classes. It's therefore useful for manipulating, and not merely storing,
 -- quantities of unknown dimension.
-newtype DynQuantity a = DynQuantity (Maybe (AnyQuantity a))
-  deriving (Eq, Data, Generic, Generic1, Typeable, Show)
+data DynQuantity a = GoodQuantity (AnyQuantity a) | BadQuantity
+  deriving (Data, Generic, Generic1, Typeable, Show)
+
+instance Eq a => Eq (DynQuantity a) where
```

Comment:
That probably is the use case for promoting only to `AnyQuantity`, if you want to perform an equality test but don't know the dimension of `Quantity` to promote to.

That said equality testing probably isn't all that useful other than for caching type checks, because most people will be using approximate arithmetic.

AIUI the `Eq` and especially `Ord` instances for `Float` and `Double` are considered a necessary misfeature because the syntactic overhead of doing something else is high, but I don't see the need here really.